tomcat-8.5/src/org/dogtagpki/tomcat/JSSContext.java

```diff
@@ -0,0 +1,102 @@
+package org.dogtagpki.tomcat;
+
+import java.util.List;
+import java.security.Provider;
+import java.security.Security;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+
+import org.apache.tomcat.util.net.SSLContext;
+
+import org.mozilla.jss.provider.javax.crypto.JSSKeyManager;
+import org.mozilla.jss.ssl.javax.JSSEngine;
+import org.mozilla.jss.ssl.javax.JSSParameters;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JSSContext implements org.apache.tomcat.util.net.SSLContext {
+    public static Logger logger = LoggerFactory.getLogger(JSSContext.class);
+
+    private javax.net.ssl.SSLContext ctx;
+    private JSSParameters params;
+
+    public JSSContext(String alias, List<String> negotiableProtocols) {
+        logger.debug("TomcatJSSContext(" + alias + ", ...");
+
+        params = new JSSParameters();
+        params.setAlias(alias);
+        params.setProtocols(negotiableProtocols.toArray(new String[0]));
+    }
+
+    public void init(javax.net.ssl.KeyManager[] kms, javax.net.ssl.TrustManager[] tms, java.security.SecureRandom sr) throws java.security.KeyManagementException {
+        logger.debug("TomcatJSSContext.init(...)");
+        try {
+            ctx = javax.net.ssl.SSLContext.getDefault();
+        } catch (java.security.NoSuchAlgorithmException e) {
+            throw new java.security.KeyManagementException(e.getMessage(), e);
+        }
+        ctx.init(kms, tms, sr);
```

Comment:
According to [SSLContext.getDefault()](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html#getDefault--):
> The default context is immediately usable and does not require initialization.

Maybe it should be something like this?
```
ctx = SSLEngine.getInstance("TLS", "Mozilla-JSS");
ctx.init(kms, tms, sr);
```

Reply:
Doesn't require initialization, but in our case, can take re-initialization. In particular, it doesn't have a `KeyManager` / `TrustManager` yet (they're `null` in the default implementation), but we have a `KeyManager` here that we want passed in. 

But yes, I'll update this to your version. Thanks!